### DISCUSSION TITLE
SSL Self-Hosting

### DISCUSSION THREAD

**Discussion by mrvillage** [supabase/supabase#3469] [Questions] [answered]
Hello! I'm setting up a self-hosted supabase instance and don't know where to grab the SSL certificate from. Would someone be able to direct me in the right direction? Thanks!

**Comment by inian** [comment, marked as answer]
Hi @mrvillage, for the self hosted setup you will need to generate a SSL cert for your domain from a CA like LetsEncrypt, Digicert and you can add the cert to Kong. If you are just need this for local development, you can create a self signed certificate. There is more information on how to do this [here](https://support.konghq.com/support/s/article/How-to-setup-Kong-to-serve-an-SSL-certificate-for-API-requests).

The process is slightly different for setting up SSL for postgres, let me know if you are looking for that

**Reply by mrvillage** [reply]
Thanks!

**Reply by rudolphmax** [reply]
@inian How would I assign my self-signed cert to the local supabase servers?

**Reply by budisetiawan760** [reply]
hi  @inian may i know to set ssl for postgres connection? i already add self cert but seems like it need to config from supavisor

**Reply by khashashin** [reply]
@budisetiawan760 I have the same question, did you solved this?

**Reply by khashashin** [reply]
@budisetiawan760  nvm, found it, you need to specify the path to the certificate and key in supavizor env:

```
environment:
  GLOBAL_DOWNSTREAM_CERT_PATH: '/etc/ssl/server.crt'
  GLOBAL_DOWNSTREAM_KEY_PATH: '/etc/ssl/server.key'
```

**Reply by khashashin** [reply]
Also don't forget to enforce the ssl in pooler.exs file:

https://supabase.github.io/supavisor/configuration/tenants/

```
enforce_ssl - enforce an SSL connection on client connections
```

**Reply by hardillb** [reply]
Sorry to jump on a old thread, but I haven't had an answer to my question (https://github.com/orgs/supabase/discussions/40246) about how to get supavisor to pick a new certificate at renewal time without a full restart.

Sending a HUP doesn't appear work (crashes the supavisor container).

**Comment by crackedbeefcake** [comment]
is it really true that in dbless mode kong only accepts ssl certificates if they are HARDCODED into the kong.yaml? 
its the only way i was able to get it to work but that seems absurdly bad.

**Comment by Stokestack** [comment]
Is there any more information available on this? I looked at the recommended Kong documentation, but I can't figure out how to follow those instructions for local Supabase development. I'm baffled by:

> 5) Upload the certificate and key to Kong. We are creating the SNI entity with the same Admin API request 
> 
> curl -k -X POST \
>   https://kong.lan:8444/certificates \
>   -H 'Content-Type: multipart/form-data' \
>   -F cert=@./kong.lan.pem \
>   -F key=@./kong.lan.key \
>   -F snis[]=kong.lan

I have no idea how to translate that to putting the certificates somewhere on my local machine.

**Reply by swingislee** [reply]
Have you solved this problem? I'm having the same confusion now

**Reply by Stokestack** [reply]
Nope. I gave up on Supabase entirely. Too much undocumented, too many roadblocks that nobody is able or willing to answer questions about. Stuff like this wasted weeks of my time. I wrote my own user-authentication and validation logic, and then got back to work on functionality. Much happier now.

**Reply by wittyphantom333** [reply]
I agree.  While this would be an awesome product to use, the fact that the "open source" version is so broken and missing features of live that its just another one of the companies out there right now that try to use the opensource community to make their product known and succeed, but in such a broken manner that you are better off using their paid for offering.  This is becoming the norm in the industry, OpenAI anyone?, and its really sad.  So many companies turning their backs on devs that promote their products its turning into a joke.

**Comment by ShunL12324** [comment]
The easiest way I think is to use Nginx as a reverse proxy:

```nginx
# In Nginx's config file
server {
    listen 9443 ssl http2;
    server_name xxxxx.com;

    # SSL Configuration
    ........
    location / {
        # Use http since Kong is not configured with SSL
        proxy_pass http://127.0.0.1:8000; 
    ........
}
```

If you use the same port as Kong (8000 & 8443), you may need to set the Kong container to only listen on 127.0.0.1:

```yaml
kong:
    .........
    ports:
      - 127.0.0.1:${KONG_HTTP_PORT}:8000/tcp
      - 127.0.0.1:${KONG_HTTPS_PORT}:8443/tcp
```

**Comment by cingh-jasdeep** [comment]
![Screenshot 2025-01-28 at 4 43 16 PM](https://github.com/user-attachments/assets/a29a2b49-9dbc-4493-8575-6a019b156221)

I got it working.

Step 1
I generated a certificate for my domain using certbot.
https://certbot.eff.org/instructions

Step 2
I followed the instructions on this post:
https://towardsdev.com/self-hosted-supabase-with-lets-encrypt-certificate-for-kong-2a6a8dd298db

and it's GitHub repo
https://github.com/Razikus/supabase-letsencrypt

**Reply by Stokestack** [reply]
Nice work! Thanks for posting this.

**Reply by Razikus** [reply]
Maybe at some point I will try to merge it to Supabase

**Reply by Inspirational-sausage-dog** [reply]
@Razikus Hello, have you encountered the problem where storage http resumable requests are sending GETs at 8443 port instead of assigned in .env file? If so, how did you bypass it. It seems that SUPABASE_PUBLIC_URL only parses the host for some reason.

**Reply by rudolphmax** [reply]
@Inspirational-sausage-dog no idea what's causing it, but a quick fix is to just set API port to `8443` in `config.toml`, then specify that in your `.env` as well.

**Reply by Razikus** [reply]
Hey guys sorry i didn't noticed. All should work normally i will soon update the thing with latest images. Thanks for reply @rudolphmax

**Reply by rudolphmax** [reply]
@Razikus have you encountered this as well? I am not even using your setup to be honest, just stumbled over someone with the same problem.

**Reply by Razikus** [reply]
> @Razikus have you encountered this as well? I am not even using your setup to be honest, just stumbled over someone with the same problem.

No i don't

**Comment by aantti** [comment]
Added a how-to guide here: https://supabase.com/docs/guides/self-hosting/self-hosted-proxy-https

**Reply by rudolphmax** [reply]
@aantti You are a hero for cleaning up the confusion and adding something proper for people to find on the website. Many thanks!

**Reply by aantti** [reply]
😅
